File: 백준/Silver/9012. 괄호/괄호.js
'use strict';
const path = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = require('fs').readFileSync(path).toString().trim().split('\n');

const [N, ...brackets] = input;
let result = [];

for (let bracket of brackets) {
  const stack = [];
  let isBalanced = true;

  for (let char of bracket) {
    if (char === '(') {
      stack.push(char);
    } else if (char === ')') {
      if (stack.length > 0 && stack[stack.length - 1] === '(') {
        stack.pop();
      } else {
        isBalanced = false;
        break;
      }
    }
  }

  isBalanced && stack.length === 0 ? result.push('YES') : result.push('NO');
}

console.log(result.join('\n'));

File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline 

N = int(input())
results = []

for _ in range(N):
  data = input().rstrip()
  stack = []
  is_valid = True
  
  for char in data:
    if char == '(':
      stack.append(char)
      
    elif char == ')':
      if not stack:
        is_valid = False
        break
      else:
        stack.pop()
        
        
  if is_valid and not stack:
    results.append('YES')
  else:
    results.append('NO')

print('\n'.join(results))